### DISCUSSION TITLE
Where to get type `referenced` from?

### DISCUSSION THREAD

**Discussion by ringods** [FormidableLabs/groqd#246] [Q&A] [answered]
Hello @scottrippey, I'm trying `groq-builder` in my site setup. I have followed the instructions using `@sanity-typed/types` to annotate my schema and create the exported type `SanityValues` in my `sanity.config.ts`.

I then run `ts-simplify` and generate the standalone Sanity types for my website project. All well so far.

The `groq-builder` README file near the end contains two typos. To import the types, the instruction currently reads:

```ts
import { createGroqBuilder, ExtractDocumentTypes } from 'groq-builder';
import { referenced, SanityValues } from './sanity-schema'; // This is the generated file
```

but should be:

```ts
import { createGroqBuilder } from 'groq-builder';
import type { ExtractDocumentTypes } from 'groq-builder';
import type { referenced, SanityValues } from './sanity-schema'; // This is the generated file
```

Aside from the missing `type` keyword, the import of `SanityValues` works, but not `referenced`. When I look at `sanity-schema.d.ts` file, the only exported value is the type `SanityValues`. Where should this `referenced` type come from?

**Comment by scottrippey** [comment, marked as answer]
@ringods Thank you for trying it out!  While this is in alpha stage, we're getting close to a beta, and having you try this out is really useful and helpful.  I'd love to help with this issue.

When the Sanity config contains any "reference" types, the generated types will include a special unique symbol called `referenced`.  For example, here's an abbreviated `Variant` document definition:
```ts
import { defineArrayMember, defineField, defineType } from "@sanity-typed/types";

export const Variant = defineType({
  name: "variant",
  type: "document",
  fields: [
    defineField({ name: "name", type: "string", }),
    defineField({ name: "price", type: "number", }),
    defineField({
      name: "flavours",
      title: "Flavours",
      type: "array",
      of: [
        defineArrayMember({
          type: "reference",
          to: [{ type: "flavour" } as const],
        }),
      ],
    }),
  ],
});
```
Notice the `flavours` field is an array of references, pointing to `flavour` documents.

After running `ts-simplify`, you ought to get a type that looks like this:
```ts
export declare const referenced: unique symbol;
export type SanityValues = {
  variant: {
    _type: "variant";
    _id: string;
    name: string;
    price: number;
    flavours: {
        _type: "reference";
        [referenced]: "flavour";
        _ref: string;
        _key: string;
    }[];
  };
};
```

Notice how the `[referenced]: "flavour"` symbol is defined in there, which is needed to be able to understand how to follow the reference.

So, hopefully that explains what to expect.  Can you please let me know what you're seeing instead?

**Reply by ringods** [reply]
@scottrippey that is not what I experience. For instance, In my Sanity Studio setup, I have a `post` document type with this field:

```
import {
    defineArrayMember,
    defineField,
    defineType,
} from "@sanity-typed/types";

...

        defineField({
            name: "authors",
            title: "Authors",
            type: "array",
            of: [
                defineArrayMember({
                    type: "reference",
                    to: [
                        { type: "author" as const }
                    ]
                }),
            ],
        }),
```

My `sanity.config.ts`:

```ts
import { defineConfig } from "@sanity-typed/types";
import type { InferSchemaValues } from "@sanity-typed/types";

import { deskTool } from 'sanity/desk'
import { visionTool } from '@sanity/vision'
import { groqdPlaygroundTool } from "groqd-playground";

import { codeInput } from '@sanity/code-input'

import { schemaTypes } from './schemas'

const config = defineConfig({
  name: 'default',
  title: 'ringo.de-smet.name',

  projectId: '12c24oc8',
  dataset: 'production',

  plugins: [
    deskTool(),
    visionTool(),
    groqdPlaygroundTool(),
    codeInput(),
  ],

  schema: {
    types: schemaTypes,
  },
})

export default config;

/** Typescript type of all types! */
export type SanityValues = InferSchemaValues<typeof config>;
```

After running this through `ts-simplify`, I get this as output for `post`:

```
post: { 
    title?: string | undefined; 
    slug?: { 
        _type: 'slug'; 
        current: string; 
    } | undefined; 
    body?: any[] | undefined; 
    _id: any; 
    _createdAt: any; 
    _updatedAt: any; 
    _rev: any; 
    mainImage?: any; 
    publishedAt?: any; 
    excerpt?: any; 
    authors?: any; 
    categories?: 
    any; _type: 
    any; 
}
```

Nowhere in the generated `sanity-schema.d.ts` file is the keyword `referenced`.

**Reply by scottrippey** [reply]
Ok, cool, making progress.  The problem is `authors?: any`, which is supposed to be more detailed than that; it should be like `authors?: { _type: "reference", [referenced]: "author" }`.
It seems like TS is having a hard time inferring that type, so it's dumping `any`.  I'd like to know what that type looks like before `ts-simplify` ... like, if you did:
```
let x: SanityValues['post']['authors'];
``` 
and then hover over `x` in your IDE ... does it just show `any` too?  Or perhaps there's an error message, or a real type?

**Reply by ringods** [reply]
@scottrippey short screenshot of the `Post` type in TS on in my Sanity Studio project:

<img width="690" alt="Screenshot 2023-12-05 at 19 35 45" src="https://github.com/FormidableLabs/groqd/assets/77923/54977758-e624-47b2-9af1-3982f30c05d5">

and `authors` property:

<img width="477" alt="Screenshot 2023-12-05 at 19 39 29" src="https://github.com/FormidableLabs/groqd/assets/77923/7f3c93dd-97fe-4349-a3d2-a629ef3b2f70">


Now, by switching from `sanity` to `@sanity-typed/types`, I noticed today that the content editor in Studio crashes. I guess there is something wrong with my schema which only now pops up by making the setup stricter. I will dig deeper and report back.

**Reply by scottrippey** [reply]
What do you see if you scroll down on the `Authors` type?  I expect to see a `_type: "reference", [referenced]: "author"` near the bottom.

**Reply by ringods** [reply]
I ended up with dependency problems in my Sanity Studio setup which I had to fix first. After getting that fixed, I restarted my attempt to enable the type generation.

I first completed my schema with the different `define...` utility methods from `sanity` package. After verifying that all still worked, I switched the import to `@sanity-typed/types` and did the `as const` changes as documented.

Again, I first checked that my Studio still ran correctly. With all in place, I added the line to export all types as `SanityValues` and used `ts-simplify` to generate the standalone typings file. 

This now contains the type `SanityValues` and the const `referenced`! 🎉

**Reply by saiichihashimoto** [reply]
If you're using sanity-typed, you _could_ import `referenced` directly from `@sanity-typed/types`. Also, if your build setup works with it, you can import your types directly from `sanity.config.ts` rather than using `ts-simplify` to build a file. `ts-simplify` is an optimization that technically isn't required. Initially `@sanity-typed/types` and `@sanity-typed/groq` were relatively slow so this helped, but their performance has far improved. I'd try without them before adding a build step. Also, the [pnpm setup breaks with the content editor](https://github.com/saiichihashimoto/sanity-typed/issues/394), which may have been your issue there.

I'm the author of sanity-typed.
